# Yoarui Guo (yg2815)

# lab 5

In [1]:
# Some preliminary set up code:
from __future__ import division
import pandas as pd
import numpy as np
import statsmodels.api as sm
import statsmodels.formula.api as smf
import os
import matplotlib.pyplot as plt

import warnings
warnings.filterwarnings("ignore")

In [2]:
os.chdir('/Users/gyr/Desktop/5019_python/dataset') # change working directory

g = pd.read_csv("GSS.2006.csv")
g.head()

vpsu  vstrat  adults  ballot  dateintv  famgen  form  formwt  gender1  \
0     1    1957       1       3       316       2     1       1        2   
1     1    1957       2       2       630       1     2       1        2   
2     1    1957       2       2       314       2     1       1        2   
3     1    1957       1       1       313       1     2       1        2   
4     1    1957       3       1       322       2     2       1        2   

   hompop  ...  away7  gender14  old14  relate14  relhh14  relhhd14  relsp14  \
0       3  ...    NaN       NaN    NaN       NaN      NaN       NaN      NaN   
1       2  ...    NaN       NaN    NaN       NaN      NaN       NaN      NaN   
2       2  ...    NaN       NaN    NaN       NaN      NaN       NaN      NaN   
3       1  ...    NaN       NaN    NaN       NaN      NaN       NaN      NaN   
4       3  ...    NaN       NaN    NaN       NaN      NaN       NaN      NaN   

   where12  where6  where7  
0      NaN     NaN     NaN  
1      NaN     NaN     NaN  
2      NaN     NaN     NaN  
3      NaN     NaN     NaN  
4      NaN     NaN     NaN  

[5 rows x 1261 columns]

# 1. Run a multiple linear probability model (have at least 2 Xs in the model). Tell me how you think your independent variables will affect your dependent variable. Interpret your results. Were your expectations correct? Why or why not?

## The question is “How often do people find their work stressful".  The answer range is from 1 to 5, which represents always to never.

## The dependent variable is 'stress', and the independent variables I have chosen are five possible influencing factors: "xhaustn","age", "babies", "prestg80","income". 

## I speculate that stress at work rises with the number of children and prestige, and decreases with age and income. Stress at work will be higher when work makes people more exhausted.

In [15]:
sub = g.dropna(subset = ["stress", "xhaustn","age", "babies", "prestg80","income","wrkstat"]) 
sub

vpsu  vstrat  adults  ballot  dateintv  famgen  form  formwt  gender1  \
15       1    1958       1       4       320       1     2       1        2   
22       2    1958       1       4       323       2     1       1        2   
23       2    1958       1       4       703       1     2       1        1   
27       1    1959       1       4       501       1     1       1        1   
32       1    1959       4       4       419       2     2       1        2   
...    ...     ...     ...     ...       ...     ...   ...     ...      ...   
4481     2    2085       2       4       620       2     1       1        2   
4494     1    2090       1       4       517       4     1       1        2   
4496     2    2090       3       4       617       1     2       1        1   
4502     2    2091       1       4       719       2     1       1        2   
4507     2    2092       2       4       709       2     1       1        2   

      hompop  ...  away7  gender14  old14  relate14  relhh14  relhhd14  \
15         1  ...    NaN       NaN    NaN       NaN      NaN       NaN   
22         6  ...    NaN       NaN    NaN       NaN      NaN       NaN   
23         1  ...    NaN       NaN    NaN       NaN      NaN       NaN   
27         1  ...    NaN       NaN    NaN       NaN      NaN       NaN   
32         6  ...    NaN       NaN    NaN       NaN      NaN       NaN   
...      ...  ...    ...       ...    ...       ...      ...       ...   
4481       4  ...    NaN       NaN    NaN       NaN      NaN       NaN   
4494       2  ...    NaN       NaN    NaN       NaN      NaN       NaN   
4496       3  ...    NaN       NaN    NaN       NaN      NaN       NaN   
4502       2  ...    NaN       NaN    NaN       NaN      NaN       NaN   
4507       3  ...    NaN       NaN    NaN       NaN      NaN       NaN   

      relsp14  where12  where6  where7  
15        NaN      NaN     NaN     NaN  
22        NaN      NaN     NaN     NaN  
23        NaN      NaN     NaN     NaN  
27        NaN      NaN     NaN     NaN  
32        NaN      NaN     NaN     NaN  
...       ...      ...     ...     ...  
4481      NaN      NaN     NaN     NaN  
4494      NaN      NaN     NaN     NaN  
4496      NaN      NaN     NaN     NaN  
4502      NaN      NaN     NaN     NaN  
4507      NaN      NaN     NaN     NaN  

[865 rows x 1261 columns]

In [13]:
sub = g.dropna(subset = ["jobsat", "hrs1", "age", "finrela", "safehlth", "proudemp"]) 
sub

Empty DataFrame
Columns: [vpsu, vstrat, adults, ballot, dateintv, famgen, form, formwt, gender1, hompop, id, intage, intid, intyrs, mode, oversamp, phase, race, reg16, region, relate1, relhh1, relhhd1, respnum, rvisitor, sampcode, sample, sex, size, spaneng, srcbelt, version, visitors, wtss, wtssall, wtssnr, xnorcsiz, year, hispanic, rplace, degree, hefinfo, wrkstat, racecen1, marital, phone, comprend, coop, feeused, hhrace, lngthinv, educ, childs, ethnum, age, cohort, intrace1, attend, inthisp, hhtype, hhtype1, earnrs, whoelse6, whoelse1, whoelse2, whoelse3, whoelse4, whoelse5, partyid, relig, teens, preteen, babies, reliten, dwelling, vote04, mar1, zodiac, old1, fund, polviews, income, income06, wrkslf, isco681, isco88, occ80, prestg80, wrkgovt, sei, intethn, intsex, indus80, eth1, trust, realinc, coninc, ethnic, health, union, ...]
Index: []

[0 rows x 1261 columns]

## I am going to recode to look variable 'stress' at if someone says "Always" and "Often" vs. everything else.

In [4]:
conditions = [
    (sub['stress'] <= 2),
    (sub['stress'] > 2)]
choices = [1, 0]
sub['str'] = np.select(conditions, choices, default=np.nan)

In [5]:
sub.str.value_counts()

0.0    530
1.0    335
Name: str, dtype: int64

## I reverse code these independent variable 'xhaustn' to make the interpretion easily.

In [6]:
## reverse code variable 'xhaustn' - higher variable means feel more exhaused from work ##
sub['rxhaustn'] = 6 - sub.xhaustn

# I will subset LPM on only full-time working people

In [7]:
lm1 = smf.ols(formula = 'str ~ rxhaustn + age + babies + prestg80 + income', 
              subset=(sub['wrkstat']==1), 
              data = sub).fit()
print (lm1.summary())

                            OLS Regression Results                            
Dep. Variable:                    str   R-squared:                       0.163
Model:                            OLS   Adj. R-squared:                  0.157
Method:                 Least Squares   F-statistic:                     26.82
Date:                Sun, 26 Jun 2022   Prob (F-statistic):           7.96e-25
Time:                        16:00:33   Log-Likelihood:                -432.28
No. Observations:                 695   AIC:                             876.6
Df Residuals:                     689   BIC:                             903.8
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -0.5092      0.170     -3.000      0.0

## This LPM result shows that the variables 'rxhaustn'  and 'prestg80' are statistically significant for the dependent variable 'stress', while the other three variables all have p-values much greater than 0.05, so they are not statistically significant to stress.

## For every category increase in feeling exhausted at work, a full-time employee is 0.206  points more likely to gain stress from work on average, net of prestg80 and other variables.. 

## Having a higher prestige, makes full-time employee 0.46 percentage points more likely to gain stress from work on average, controlling for level of feeling exhausted at work and other variables.

## This result is roughly in line with my expectations. People are more likely to be more stressed at work when they feel more exhausted at work, or are in positions of high prestige.

# 2. Run a multiple (binary) logistic model. (It can be the same as the above LPM or a new model.) If it is a new model, tell me how you think your independent variables will affect your dependent variable. Interpret your results in the logit scale. Were your expectations correct? Why or why not?

In [8]:
logit1 = sm.formula.logit(formula = 'str ~ rxhaustn + age + babies + prestg80 + income', 
                          subset=(sub['wrkstat']==1), 
                          data = sub).fit()
print (logit1.summary()) 

Optimization terminated successfully.
         Current function value: 0.590676
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:                    str   No. Observations:                  695
Model:                          Logit   Df Residuals:                      689
Method:                           MLE   Df Model:                            5
Date:                Sun, 26 Jun 2022   Pseudo R-squ.:                  0.1293
Time:                        16:00:33   Log-Likelihood:                -410.52
converged:                       True   LL-Null:                       -471.50
Covariance Type:            nonrobust   LLR p-value:                 1.207e-24
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -4.8973      0.887     -5.520      0.000      -6.636      -3.159
rxhaustn       0.9886      0.

## From logit model result, for every category increase in feeling exhausted at work, a full time employee increases his/her logit by 0.989 of having strees from work on average, net of prestg80 and other variables.

## Having higher prestige increases a full time employee’s log-odds on having strees from work by 0.022 on average, controlling for level of feeling exhausted at work and other variables.

# 3. Get odds ratios from your logit model in Question 2 and interpret some of them.

In [9]:
np.exp(logit1.params)

Intercept    0.007466
rxhaustn     2.687346
age          1.005737
babies       1.200495
prestg80     1.022243
income       0.986019
dtype: float64

## For every category increase in feeling exhausted at work, a full time employee increases his/her odds by 1.69 of having strees from work, net of prestg80 and other variables.

## Having higher prestige increases a full time employee’s odds on having strees from work by 2.22%, controlling for level of feeling exhausted at work and other variables.

# 4. Extra Credit: Get predicted probabilities from your logit model in Question 2 for some constellations of X values and interpret the results.

In [10]:
def logit2prob (logit):
    odds = np.exp(logit)
    prob = odds / (1 + odds) 
    return(prob);


intercept = logit1.params.Intercept
b_rxhaustn = logit1.params.rxhaustn
b_age = logit1.params.age
b_babies = logit1.params.babies
b_prestg80 = logit1.params.prestg80
b_income = logit1.params.income

## I want to know if I will feel stressful about my job after graduation. I am going to find out the predicted probability of being stressed for a person who feel moderately exhausted from work, is 24 years old, has a relatively high prestige job of 50, no baby, and can earn the average income.

In [12]:
logits_str = intercept + (3 * b_rxhaustn) + (24 * b_age) + (50 * b_prestg80) + (0 * b_babies) \
             + (sub.income.mean() * b_income)
logit2prob(logits_str)

0.29849783655659134

## I see my self has a 29.85 probability of being stressed from work. It looks like I don't have to wait until graduation, and this result is starting to make me feel anxious and stressed now.